control-plane-operator/controllers/hostedcontrolplane/routecm/reconcile.go

```diff
@@ -0,0 +1,239 @@
+
+type RouteControllerManagerReconciler struct {
+}
+
+func NewComponent() component.ControlPlaneComponent {
+	return &component.ControlPlaneWorkload{
+		DeploymentReconciler: &RouteControllerManagerReconciler{},
+		ResourcesReconcilers: []component.GenericReconciler{
+			component.NewReconcilerFor(&corev1.ConfigMap{}).
+				WithName(ConfigMapName).
+				WithReconcileFunction(ReconcileOpenShiftRouteControllerManagerConfig).
+				Build(),
+			component.NewReconcilerFor(&corev1.Service{}).
+				WithName(ComponentName).
+				WithReconcileFunction(ReconcileService).
+				WithPredicate(component.DisableIfAnnotationExist(hyperv1.DisableMonitoringServices)).
+				Build(),
+			component.NewReconcilerFor(&prometheusoperatorv1.ServiceMonitor{}).
+				WithName(ComponentName).
+				WithReconcileFunction(ReconcileServiceMonitor).
+				WithPredicate(component.DisableIfAnnotationExist(hyperv1.DisableMonitoringServices)).
+				Build(),
+		},
+
+		MultiZoneSpreadLabels: openShiftRouteControllerManagerLabels(),
+	}
+}
+
+// Name implements controlplanecomponent.DeploymentReconciler.
+func (r *RouteControllerManagerReconciler) Name() string {
+	return ComponentName
+}
+
+// ReconcileDeployment implements controlplanecomponent.DeploymentReconciler.
+func (r *RouteControllerManagerReconciler) ReconcileDeployment(cpContext component.ControlPlaneContext, deployment *appsv1.Deployment) error {
+	image := cpContext.ReleaseImageProvider.GetImage("route-controller-manager")
+
+	config := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ConfigMapName,
+			Namespace: cpContext.HCP.Namespace,
+		},
+	}
+	if err := cpContext.Client.Get(cpContext, client.ObjectKeyFromObject(config), config); err != nil {
+		return fmt.Errorf("failed to get openshift controller manager config: %w", err)
+	}
+
+	configBytes, ok := config.Data[configKey]
+	if !ok {
+		return fmt.Errorf("openshift controller manager configuration is not expected to be empty")
+	}
+	configHash := util.ComputeHash(configBytes)
+
+	maxSurge := intstr.FromInt(0)
+	maxUnavailable := intstr.FromInt(1)
+	deployment.Spec.Strategy.Type = appsv1.RollingUpdateDeploymentStrategyType
+	deployment.Spec.Strategy.RollingUpdate = &appsv1.RollingUpdateDeployment{
+		MaxSurge:       &maxSurge,
+		MaxUnavailable: &maxUnavailable,
+	}
+	if deployment.Spec.Selector == nil {
+		deployment.Spec.Selector = &metav1.LabelSelector{
+			MatchLabels: openShiftRouteControllerManagerLabels(),
+		}
+	}
+	deployment.Spec.Template.ObjectMeta.Labels = openShiftRouteControllerManagerLabels()
+	deployment.Spec.Template.ObjectMeta.Annotations = map[string]string{
+		configHashAnnotation: configHash,
+	}
+	deployment.Spec.Template.Spec.AutomountServiceAccountToken = ptr.To(false)
+	deployment.Spec.Template.Spec.Containers = []corev1.Container{
+		util.BuildContainer(routeOCMContainerMain(), buildRouteOCMContainerMain(image)),
+	}
+
+	return nil
+}
+
+var (
+	volumes = component.Volumes{
```

Comment:
If we can figure out a way to not have these be package level vars, that'd be awesome. For example, this can be inlined in the Volumes function below, and then when containers need to be built/reconciled, we automatically pass the volumes field to them.

Reply:
yeah that would be better for sure. I tried to do that, but it turns out the configMap reconciler also needs the volumes.
https://github.com/openshift/hypershift/blob/5479c0ca8b0b59f0b9561a25d0955c293e4f47f6/control-plane-operator/controllers/hostedcontrolplane/routecm/config.go#L56

I have found the same pattern in multiple other components as well. we could pass the volumes to the genricReconciler reconciler function as well, but I am not sure if that makes sense.

Reply:
> we could pass the volumes to the genricReconciler reconcile function too

I think that would work